ng-dev/pr/merge/strategies/api-merge.ts

```diff
@@ -48,23 +58,55 @@ export class GithubApiMergeStrategy extends AutosquashMergeStrategy {
    */
   override async merge(pullRequest: PullRequest): Promise<void> {
     const {githubTargetBranch, prNumber, needsCommitMessageFixup, targetBranches} = pullRequest;
-    const method = this.getMergeActionFromPullRequest(pullRequest);
     const cherryPickTargetBranches = targetBranches.filter((b) => b !== githubTargetBranch);
-
-    // Squash and Merge will create a single commit message and thus we can use the API to merge.
-    if (
-      method === 'rebase-with-fixup' &&
-      (pullRequest.needsCommitMessageFixup || (await this.hasFixupOrSquashCommits(pullRequest)))
-    ) {
-      return super.merge(pullRequest);
-    }
-
+    const commits = await this.getPullRequestCommits(pullRequest);
+    const {squashCount, fixupCount, normalCommitsCount} = await this.getCommitsInfo(pullRequest);
+    const method = this.getMergeActionFromPullRequest(pullRequest);
+    let pullRequestCommitCount = pullRequest.commitCount;
     const mergeOptions: OctokitMergeParams = {
       pull_number: prNumber,
-      merge_method: method === 'rebase-with-fixup' ? 'rebase' : method,
+      merge_method: method === 'auto' ? 'rebase' : method,
       ...this.git.remoteParams,
     };
 
+    // When the merge method is `auto`, the merge strategy will determine the best merge method
+    // based on the pull request's commits.
+    if (method === 'auto') {
+      const hasFixUpOrSquashAndMultipleCommits =
+        normalCommitsCount > 1 && (fixupCount > 0 || squashCount > 0);
+
+      // If the PR has fixup/squash commits against multiple normal commits, or if the
+      // commit message needs to be fixed up, delegate to the autosquash merge strategy.
+      if (needsCommitMessageFixup || hasFixUpOrSquashAndMultipleCommits) {
+        return super.merge(pullRequest);
+      }
+
+      const hasOnlyFixUpForOneCommit =
+        normalCommitsCount === 1 && fixupCount > 0 && squashCount === 0;
```

Comment:
What if someone accidentally has fixup commits for a commit outside the PR?

Reply:
That is handled by the commit validator lint check right?